File: generated-code/rust/codecraft/TcpReadWrite/src/color.rs
use super::*;

/// RGBA Color
#[derive(Clone, Debug)]
pub struct Color {
    /// Red component
    pub r: f32,
    /// Green component
    pub g: f32,
    /// Blue component
    pub b: f32,
    /// Alpha (opacity) component
    pub a: f32,
}

impl trans::Trans for Color {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.r.write_to(writer)?;
        self.g.write_to(writer)?;
        self.b.write_to(writer)?;
        self.a.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let r: f32 = trans::Trans::read_from(reader)?;
        let g: f32 = trans::Trans::read_from(reader)?;
        let b: f32 = trans::Trans::read_from(reader)?;
        let a: f32 = trans::Trans::read_from(reader)?;
        Ok(Self {
            r,
            g,
            b,
            a,
        })
    }
}

File: generated-code/pascal/codecraft/TcpReadWrite/Color.pas
unit Color;

{$mode delphi}{$H+}

interface

uses
    Stream,
    SysUtils;

type
    // RGBA Color
    TColor = class
        // Red component
        r: Single;
        // Green component
        g: Single;
        // Blue component
        b: Single;
        // Alpha (opacity) component
        a: Single;
        constructor Create(r: Single; g: Single; b: Single; a: Single);
        // Read Color from input stream
        class function ReadFrom(stream: TStream): TColor; static;
        // Write Color to output stream
        procedure WriteTo(stream: TStream);
        function ToString: ansistring; override;
    end;

implementation

constructor TColor.Create(r: Single; g: Single; b: Single; a: Single);
begin
    self.r := r;
    self.g := g;
    self.b := b;
    self.a := a;
end;

class function TColor.ReadFrom(stream: TStream): TColor;
var a: Single;
var b: Single;
var g: Single;
var r: Single;
begin
    r := stream.ReadSingle;
    g := stream.ReadSingle;
    b := stream.ReadSingle;
    a := stream.ReadSingle;
    result := TColor.Create(r, g, b, a);
end;

procedure TColor.WriteTo(stream: TStream);
begin
    stream.WriteSingle(r);
    stream.WriteSingle(g);
    stream.WriteSingle(b);
    stream.WriteSingle(a);
end;

function TColor.ToString: ansistring;
begin
    result := 'Color {';
    result += 'r=';
    result += FloatToStr(r);
    result += ', ';  
    result += 'g=';
    result += FloatToStr(g);
    result += ', ';  
    result += 'b=';
    result += FloatToStr(b);
    result += ', ';  
    result += 'a=';
    result += FloatToStr(a);
    result += '}';
end;

end.